src/buffer.rs

```diff
@@ -100,8 +101,8 @@ impl Buffer {
                     std::mem::size_of_val(wrapper.inner_mut()),
                 ),
                 SliceParameter::H264(ref mut wrapper) => (
-                    wrapper.inner_mut() as *mut _ as *mut std::ffi::c_void,
-                    std::mem::size_of_val(wrapper.inner_mut()),
+                    wrapper.inner_mut().as_mut_ptr() as *mut std::ffi::c_void,
+                    std::mem::size_of::<bindings::VASliceParameterBufferH264>(),
```

Comment:
Mmm, I've been looking at the other branches of this match statement and think it may be worth applying this change everywhere since the type and size are always known AFAICT - WDYT? Not asking to do it here, but I suspect that more generally we could replace this boilerplate code with something a bit more concise, maybe a macro or something...

Reply:
> ... applying this change everywhere since the type and size are always known AFAICT - WDYT?

Yeah, except the obvious SliceData/EncCodedBuf

> but I suspect that more generally we could replace this boilerplate code with something a bit more concise, maybe a macro or something...

I don't want to sound too overconfident, but to be completely honest, we could look into rewriting this 😅 😅 😅  We could use a private trait in favor of `BufferType` enum. Something like:  
```rust
trait BufferElement: Into<Self::RawType> {
    const TYPE: bindings::VABufferType::Type;
    type RawType;

    fn raw_element_size() -> usize {
        std::mem::size_of::<Self::RawType>()
    }
}
```
Replace tuple structs with strongly typed and documented copy of raw struct. It would have to be converted to raw when added to the picture (or prior `vaRenderPicture`), but during this step we could clone `Rc<>`s to dependencies (like surfaces and EncCodedBuffer) to `Picture<>` so those wouldn't get dropped before sync. It would clean up some cros-codecs code.

Also `Picture::add_buffer` could accept a trait that would wrap our use cases. Like so: 

```rust
trait BufferContents {
    const TYPE: bindings::VABufferType::Type;

    fn element_size(&self) -> usize;

    fn num_elements(&self) -> usize;

    unsafe fn data_ptr(&mut self) -> *mut std::ffi::c_void;
}

// Or just use slice somehow
pub struct BufferArray<T>
where
    T: BufferElement,
{
    array: Vec<T::RawType>,
    // ... deps
}

impl<T> BufferContents for BufferArray<T>
where
    T: BufferElement,
{
 // ...
}

impl<T> BufferContents for Box<T> // Or just T
where
    T: BufferElement,
{
 // ...
}

impl BufferContents for SliceData
{
 // ...
}
```
In this way we could handle buffer arrays and single element buffers transparently. 
On top of that we maybe could look into doing some validations.

However this might be unnecessary overengineering 😅

Reply:
> However this might be unnecessary overengineering 😅

No, I think it would be very nice (and less error-prone) to have. This code has been written at the very beginning of the project, when we weren't familiar with Rust or experienced enough with VAAPI to understand how things should be done. This crate is probably full of things that would benefit from such a rewrite.

Reply:
Created https://github.com/chromeos/cros-libva/issues/11 to track this.